subsys/net/ip/net_tls.c

```diff
@@ -0,0 +1,735 @@
+	 */
+	if (context->tls->ssl.state == MBEDTLS_SSL_HANDSHAKE_OVER) {
+		len = net_pkt_appdatalen(pkt);
+
+		decrypted_pkt = net_pkt_clone(pkt, 0, NET_PKT_CLONE_HDR);
+		if (!decrypted_pkt) {
+			return;
+		}
+
+		/* Copy IP and transport header */
+		u16_t header_len = net_pkt_appdata(pkt) - pkt->frags->data;
+
+		ret = net_frag_linear_copy(decrypted_pkt->frags, pkt->frags,
+					   0, header_len);
+		if (ret < 0) {
+			goto err;
+		}
+
+		/* Set appdata and appdatalen */
+		net_pkt_set_appdata(decrypted_pkt,
+				    decrypted_pkt->frags->data + header_len);
+		net_pkt_set_appdatalen(decrypted_pkt, 0);
+
+		while (1) {
+			ret = mbedtls_ssl_read(&context->tls->ssl,
+					       context->tls->rx_ssl_buf,
+					       sizeof(context->tls->rx_ssl_buf));
+			if (ret == 0 || ret == MBEDTLS_ERR_SSL_WANT_READ ||
+				ret == MBEDTLS_ERR_SSL_WANT_WRITE) {
+				ret = 0;
+
+				break;
+			}
+
+			if (ret > 0) {
+				ret = net_pkt_append(decrypted_pkt, ret,
+						     context->tls->rx_ssl_buf,
+						     TIMEOUT_TLS_RX_MS);
+				/* Update appdatalen */
+				net_pkt_set_appdatalen(decrypted_pkt,
+						       net_pkt_appdatalen(decrypted_pkt) + ret);
+				continue;
+			}
+
+			ret = -EBADF;
+			break;
+		}
+
+		if (ret >= 0 && context->tls->tls_cb) {
+			context->tls->tls_cb(context, decrypted_pkt,
+					     status, user_data);
+		} else {
+			goto err;
+		}
+	}
+
+	return;
+
+err:
+	net_pkt_unref(decrypted_pkt);
+}
+
+static int tls_add_ca_certificate(struct net_context *context,
+				  struct net_tls_credential *ca_cert)
+{
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+	int err = mbedtls_x509_crt_parse(&context->tls->ca_chain,
+					 ca_cert->buf, ca_cert->len);
+	if (err != 0) {
+		return -EINVAL;
+	}
+#endif /* MBEDTLS_X509_CRT_PARSE_C */
+
+	return 0;
+}
+
+static void tls_set_ca_chain(struct net_context *context)
+{
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+	mbedtls_ssl_conf_ca_chain(&context->tls->config,
+				  &context->tls->ca_chain, NULL);
+
+	mbedtls_ssl_conf_authmode(&context->tls->config,
+				  MBEDTLS_SSL_VERIFY_REQUIRED);
+
+	mbedtls_ssl_conf_cert_profile(&context->tls->config,
+				      &mbedtls_x509_crt_profile_default);
+#endif /* MBEDTLS_X509_CRT_PARSE_C */
+}
+
+static int tls_set_psk(struct net_context *context,
+		       struct net_tls_credential *psk,
+		       struct net_tls_credential *psk_id)
+{
+#if defined(MBEDTLS_KEY_EXCHANGE__SOME__PSK_ENABLED)
+	mbedtls_ssl_conf_psk(&context->tls->config,
+			     psk->buf, psk->len,
+			     (const unsigned char *)psk_id->buf,
+			     psk_id->len - 1);
+#endif
+
+	return 0;
+}
+
+static int tls_mbedtls_set_credentials(struct net_context *context)
+{
+	struct net_tls_credential *credential;
+	sec_tag_t tag;
+	int i, err;
+	bool ca_cert_present = false;
+
+	for (i = 0; i < context->tls->options.sec_tag_list.sec_tag_count; i++) {
+		tag = context->tls->options.sec_tag_list.sec_tags[i];
+		credential = NULL;
+
+		while ((credential = credential_next_get(tag, credential))
```

Comment:
Iterators are a nice idiom ... for C++. I doubt they're used anywhere else in Zephyr code, and thus doubt they should be used here either.